### PR TITLE
Strategy `apply-equation`.

### PR DESCRIPTION
**Pull request by h0nzZik**

### CODE REVIEW

File: prover/strategies/matching.md [hunk 1]
```diff
@@ -177,6 +195,279 @@ Recurse over assoc-only constructors (including `pto`):
...
+
+  // \not(_) mismatched
+  rule #matchAssoc( terms:     T, _
+                  , pattern:   \not(_), _
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
+    => #matchFailure("\\not(_) does not match"), .MatchResults
+    requires \not(...) :/=K T
```
Comment (nishantjr) [inline prover/strategies/matching.md]: I'm don't think this is sound.

Comment (h0nzZik) [reply prover/strategies/matching.md]: Do you mean this general pattern that is used also for \equals and others, or do you mean this particular instance? Why it is not sound?

Comment (h0nzZik) [reply prover/strategies/matching.md]: Do you mean the general pattern that is used also for \equals and others, or do you mean this particular instance? Why it is not sound?

File: prover/strategies/matching.md [hunk 2]
```diff
@@ -177,6 +195,279 @@ Recurse over assoc-only constructors (including `pto`):
...
+
+  // \implies(_,_) matched
+  rule #matchAssoc( terms:     \implies(T1, T2), Ts
+                               => T1 ++Patterns T2 ++Patterns Ts
+                  , pattern:   \implies(P1, P2), Ps
+                               => P1 ++Patterns P2 ++Patterns Ps
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
```
Comment (nishantjr) [inline prover/strategies/matching.md]: Lets discuss what we want to do with patterns that are not constructors (possibly modulo AC).

e.g. for a term `T` to match pattern `A and B`, `T` needs to match both `A` and `B` with some constraints on the substitutions.

What is your use case for all these cases of matching?

File: prover/strategies/matching.md (hunk 1 above)
Comment (nishantjr) [reply prover/strategies/matching.md]: So when we implemented this, we only considered constructors possibly modulo AC. I'm not sure if generalizing #match so much makes sense. It becomes pretty different from what one would expect a matching function to do

Comment (nishantjr) [reply prover/strategies/matching.md]: Perhaps we want to simplify the pattern to remove instances of and, or, not etc before calling #match?

File: prover/strategies/matching.md (hunk 2 above)
Comment (h0nzZik) [reply prover/strategies/matching.md]: The use-case is the `rewrite` strategy: given an equality `A=B` (where A,B are patterns), and the goal has shape `C(A')` for some context `C` and pattern `A'` that is matched by `A`, rewrite `C(A')` to `C(B')`.

Comment (h0nzZik) [reply prover/strategies/matching.md]: By matching here, I mean syntactical matching: e.g., `\not(\not(A))` is not an instance of `A`.

Comment (nishantjr) [reply prover/strategies/matching.md]: Equality, or rewrite axiom? Do you expect `A` to have quantifiers, negations, \and, \or etc?
I'd rather have a check `isConstructorPattern` before calling the `#match` function, than implementing all these cases in `#match`

Comment (nishantjr) [reply prover/strategies/matching.md]: Would simply causing `#match` to fail for any non-constructor pattern also work?

Comment (h0nzZik) [reply prover/strategies/matching.md]: I do not understand this question:
> Equality, or rewrite axiom?

I use axioms (or claims) whose conclusion is an equality
An example is the following axiom `f_0_X`
```
symbol f(Int, Int) : Int
axiom f_0_X: \forall{X{Int}}
  \implies(
    gte(X{Int}, -1),
    \equals(f(0, X{Int}), X{Int})
  )

claim \equals(f(0,1),1)
strategy rewrite -> f_0_X
```
Here I want to pattern match `f(0, X)` against `f(0,1)`. But I also want to support more complicated things, like the following (not tested):
```
axiom A1: \forall x. f(x) = \exists y. y /\ not (y < x)
axiom A2: \forall x. g(f(x))) = x
claim: g(\exists z. z /\ not(z < 3)) = 3
strategy rewrite <- A1 // rewrites goal to g(f(3)) = 3
            . rewrite -> A2 // rewrites to 3 = 3
```

Comment (h0nzZik) [reply prover/strategies/matching.md]: > Would simply causing `#match` to fail for any non-constructor pattern also work?

No, that would not work for my use-case.

Comment (h0nzZik) [reply prover/strategies/matching.md]: So maybe rename this variant of matching to `#syntacticalMatch`? I want it to work on arbitrary formulas.

File: prover/strategies/matching.md (hunk 1 above)
Comment (h0nzZik) [reply prover/strategies/matching.md]: So maybe having a separate version `#syntacticalMatch` would resolve the issue?

**Review by nishantjr — COMMENTED** [review]
The visitor code looks great! Lets move it to `lang/kore-visitors.md` instead since it is not a strategy?

**Comment by h0nzZik** [comment]
Maybe we should also move `heatcool.md` to `lang`.

**Comment by h0nzZik** [comment]
@nishantjr I consider this finished - please review. I moved all non-strategies to `utils/`. The module `PATTERN-LENGTH` serves just as an example of a visitor - feel free to remove it even without notifying me.

**Pull request merged**